module/apmotel/span.go

```diff
@@ -0,0 +1,330 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package apmotel // import "go.elastic.co/apm/module/apmotel/v2"
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	"go.opentelemetry.io/otel/trace"
+
+	"go.elastic.co/apm/module/apmhttp/v2"
+	"go.elastic.co/apm/v2"
+)
+
+type event struct {
+	Name       string
+	Attributes []attribute.KeyValue
+	Time       time.Time
+}
+
+type status struct {
+	Code        codes.Code
+	Description string
+}
+
+type span struct {
+	mu sync.Mutex
+
+	provider *tracerProvider
+
+	startTime   time.Time
+	attributes  []attribute.KeyValue
+	events      []event
+	spanContext trace.SpanContext
+	status      status
+
+	tx   *apm.Transaction
+	span *apm.Span
+}
+
+func (s *span) End(options ...trace.SpanEndOption) {
+	config := trace.NewSpanEndConfig(options...)
+
+	if !config.Timestamp().IsZero() {
+		duration := config.Timestamp().Sub(s.startTime)
+		if s.span != nil {
+			s.span.Duration = duration
+		} else {
+			s.tx.Duration = duration
+		}
+	}
+
+	if s.span != nil {
+		s.setSpanAttributes()
+		s.span.End()
+		return
+	}
+
+	s.setTransactionAttributes()
+	s.tx.End()
+}
+
+func (s *span) AddEvent(name string, opts ...trace.EventOption) {
+	c := trace.NewEventConfig(opts...)
+	e := event{Name: name, Attributes: c.Attributes(), Time: c.Timestamp()}
+
+	s.mu.Lock()
+	s.events = append(s.events, e)
+	s.mu.Unlock()
+}
+
+func (s *span) IsRecording() bool {
+	if s.span != nil {
+		return !s.span.Dropped()
+	}
+
+	return true
```

Comment:
If the span is nil, we should check if the tx is sampled (recording)
```suggestion
	return s.tx.Sampled()
```

Reply:
Ah, I missed that method. Thanks.